enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,270 @@
+
+## Design Details
+
+### Open Questions
+
+- How will OLM rejecting an upgrade be surfaced in both the UI and the CLI?
+
+### Test Plan
+
+#### Proposed Unit Tests
+
+- Logic used to determine upgrade safety should be thoroughly tested
+
+#### Proposed E2E Tests
+
+- The cluster upgrade is not blocked by CSVs that do not report a maxOpenShiftVersion.
+- The cluster upgrade is not blocked by CSVs that report an invalid maxOpenShiftVersion.
+- The cluster upgrade is not blocked by CSVs that report a maxOpenShiftVersion greater than the current Minor OpenShift version.
+- The cluster upgrade is blocked by CSVs that report a maxOpenShiftVersion less than or equal to the current OpenShift version.
+
+### Graduation Criteria
+
+The goal of this enhancement is to provide Cluster Admins some level of assurance that service made available via OLM Managed Operators continue to remain available on a targeted OpenShift Cluster Version.
+
+This feature will initially be introduced as a Generally Available Feature, but upgrade compatibility checks may be added overtime.
+
+Proposed GA Features:
+
+- OLM determines upgrade safety based on CSV annotations
+- Console highlights upgrade safety in the UI
+- User facing documentation available at olm.operatorframework.io
+- Comprehensive unit tests exist
+- Comprehensive e2e tests exist
+
+#### Removing a deprecated feature
+
+Not applicable.
+
+#### Upgrade / Downgrade Strategy
+
+Not applicable.
+
+#### Version Skew Strategy
+
+The proposed feature should only require integration with the CVO component. All other components relevant to this enhancement proposal (OLM Managed Operators) are managed by OLM.
+
+## Implementation History
+
+- Initial enhancement proposal created.
+
+## Drawbacks
+
+- OpenShift users might be upset when a Cluster Upgrade fails if they are unfamiliar with OLM. These customers may ask why a Cluster Upgrade is rejected by the presence of an operator. Typically, forcing an upgrade is not fully supported and these customers may blame OLM for preventing a conventional upgrade.
+- There are instances where an Operator might not be supported on the targeted version that the operator is being upgraded to. If the customer is unable to remove the existing operator, this may place them in a position where they must choose between upgrading the cluster or removing the operator. This is arguably better than upgrading the cluster only to find out that a core service is no longer available on cluster, but I suspect that multiple tickets will be opened against OLM and the operators it installs when the customer finds themselves unable to upgrade.
+
+## Possible Future Work
+
+### Determine upgrade safety using OLM Managed Operator RBAC
+
+In addition to giving Operator Authors the ability to provide the Max OpenShift Version, OLM could determine upgrade readiness based on default OpenShift GVKs and those required by installed Operators. At a high level, the following workflow could be implemented:
```

Comment:
Something like this exists today, where operators which will remove a particular API in the next OpenShift minor will set `Upgradeable=False` if they see any current consumers of that API.  If the only compat issue you're worried about is Kube-objects, maybe we don't need this enhancement?  You would need this enhancement for non-Kube things like RHCOS-kernel compat (#357), so if you grow OLM-side Kube API checks in the future, there should be ways for OLM-child operators to opt-into declaring additional restrictions beyond the ones the Kube API checks turn up.

Reply:
Thanks for the review @wking!

> Something like this exists today, where operators which will remove a particular API in the next OpenShift minor will set Upgradeable=False if they see any current consumers of that API. If the only compat issue you're worried about is Kube-objects, maybe we don't need this enhancement? You would need this enhancement for non-Kube things like RHCOS-kernel compat (#357), so if you grow OLM-side Kube API checks in the future, there should be ways for OLM-child operators to opt-into declaring additional restrictions beyond the ones the Kube API checks turn up.

The existing enhancement is focused on providing operator authors or index owners a way to specify a maximum OpenShift version per bundle and is required to prevent an cluster upgrade form disabling a service provided by an operator. Based on the frequency of conversations on this topic with OLM customers this feature is highly requested.

The possible expansion of this feature to have OLM to inspect RBAC required by installed operators is simply a nice to have that allows OLM to derive when a cluster upgrade may disable services offered by operators due to the deprecation of GVKs required by the operators.

> And there's also a chance that an OLM-child operator isn't using a particular, vulnerable Kube API right now, but will try to start using it once a user touches some knob. So there is a benefit to declaring a complete set of possible consumed APIs, instead of relying entirely on current-use detection.

In general, the RBAC required by an operator is declared in the `permission` and `clusterPermission` fields of its CSV. Its possible that a knob exists which causes the operator to perform privilege escalation if it has role/rolebinding rbac, but that seems like an abusive usecase.